### DISCUSSION TITLE
Display sensor data on webpage and oled display at the same time

### DISCUSSION THREAD

**Discussion by adysanro** [micropython/micropython#9409] [Hardware & Peripherals] [answered]
Hi,
I am new to Python, HTML, boards and sensors. Using a Pi Pico W with a BME280 sensor, a SGP40 sensor and a SSD1306 oled display. Everything works fine. I am able to display sensor data on the oled and on a webpage separately. But NOT both at the same time.
So basically I want to be able to serve my data to a webpage and the display at the same time. I have managed to get this to work but only in the following scenario:
- the oled data will be displayed and refreshed as long as the webpage is displayed in the browser. The moment I close the browser window the oled data freezes with the last value. I assume this is happening because the oled display function is inside the while loop for starting the web server

I have tried putting the oled and webpage display functions in separate while loops with the use of different threads as described here by Andreas Spiess: 
https://www.youtube.com/watch?v=9vvobRfFOwk&ab_channel=AndreasSpiess

Any ideas if what I am trying to achieve is possible? (Bad) code attached

[code.zip](https://github.com/micropython/micropython/files/9640362/code.zip)

Thanks,
Adrian

**Comment by jimmo** [comment, marked as answer]
Yes, that's right. Because the "accept" function will block, the rest of your code won't run.

The best way to do what you want is with asyncio, and have separate tasks for the OLED, sensors, and web server.

I recommend using a HTTP server library that takes care of all this for you -- e.g. https://github.com/miguelgrinberg/microdot

**Reply by jimmo** [reply]
And see the general asyncio tutorial here: https://github.com/peterhinch/micropython-async/blob/master/v3/README.md

**Reply by adysanro** [reply]
Thanks! I'll have a look into this.
